<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">

  <title><![CDATA[Category: debugging | Tushar's Blog]]></title>
  <link href="http://www.tusharsaxena.com/blog/categories/debugging/atom.xml" rel="self"/>
  <link href="http://www.tusharsaxena.com/"/>
  <updated>2016-03-26T22:30:41-04:00</updated>
  <id>http://www.tusharsaxena.com/</id>
  <author>
    <name><![CDATA[Tushar Saxena]]></name>
    
  </author>
  <generator uri="http://octopress.org/">Octopress</generator>

  
  <entry>
    <title type="html"><![CDATA[Redirect_to and preservation of errors]]></title>
    <link href="http://www.tusharsaxena.com/blog/2013/03/25/redirect-to-and-preservation-of-errors/"/>
    <updated>2013-03-25T00:38:00-04:00</updated>
    <id>http://www.tusharsaxena.com/blog/2013/03/25/redirect-to-and-preservation-of-errors</id>
    <content type="html"><![CDATA[<p>Working on my side project, <a href="https://github.com/tsax/pecunia-nunc/">Pecunia-Nunc</a>, I was baffled when the errors on the object creation form would not display upon submission. The controller looked like this</p>

<p>{% codeblock lang:ruby%}
class SubscribersController &lt; ApplicationController</p>

<p>def new</p>

<pre><code>require 'digest'
@subscriber = Subscriber.new(params[:subscriber])
@subscriber.token = Digest::SHA1.hexdigest([Time.now, rand].join)
if @subscriber.save
    flash[:notice] = "Thanks for signing-up. Please confirm your email address through the email that'll be in your inbox shortly!"

    redirect_to home_path
else
    flash[:notice] = "There were errors! Please resubmit after making corrections."
    redirect_to home_path
end 
</code></pre>

<p>end
...
{% endcodeblock %}</p>

<p>The view :
{% codeblock lang:ruby%}
  &lt;% if @subscriber.errors.any? %>
  <div id="error_explanation"></p>

<pre><code>&lt;div class="alert alert-error"&gt;
  The form contains &lt;%= pluralize(@subscriber.errors.count, "error") %&gt;.
&lt;/div&gt;
&lt;ul&gt;
&lt;% @subscriber.errors.full_messages.each do |msg| %&gt;
  &lt;li&gt;* &lt;%= msg %&gt;&lt;/li&gt;
&lt;% end %&gt;
&lt;/ul&gt;
</code></pre>

<p>  </div>
&lt;% end %>
{% endcodeblock %}</p>

<p>As you can tell, the problem was not in the view at all, but in this line <code>redirect_to home_path</code>. Redirection clears the object.errors field, and therefore nothing was being displayed. I stupidly spent a lot of time fiddling with the view, instead of inspecting the controller action. Replacing the problematic line with <code>render 'new'</code> ('new' is the action/view that renders the form) solved the problem by preserving the .errors object.</p>
]]></content>
  </entry>
  
</feed>
